File: leetcode4/GCDofOddandEvenSums/a.swift
/* Copyright (C) 2025 by iamslash */

import Foundation

// 0ms 100.00% 19.62MB 100.00%
// math
// O(1) O(1)
class Solution {
    private func gcd(_ a: Int, _ b: Int) -> Int {
        if b == 0 {
            return abs(a)
        }
        return gcd(b, a % b)      
    }
    
    func gcdOfOddEvenSums(_ n: Int) -> Int {
        let nn = n + n
        let sum = nn * (nn + 1) / 2
        let sumOdd = (sum - n) / 2
        let sumEvn = sumOdd + n

        return gcd(sumEvn, sumOdd)
    }
}

File: leetcode4/GCDofOddandEvenSums/a.rs
struct Solution;

// 0ms 100.00% 2.03MB 100.00%
// math
// O(1) O(1)
impl Solution {
    fn gcd(a: i32, b: i32) -> i32 {
        if b == 0 {
            return a.abs();
        }
        return Solution::gcd(b, a % b);
    }    

    pub fn gcd_of_odd_even_sums(n: i32) -> i32 {
        let nn = n + n;
        let sum = nn * (nn + 1) / 2;
        let sum_odd = (sum - n) / 2;
        let sum_evn = sum_odd + n;

        return Solution::gcd(sum_evn, sum_odd);
    }
}

fn main() {
    println!("{}", Solution::gcd_of_odd_even_sums(1));
}